src/backend/oneapi/solve.cpp

```diff
@@ -11,111 +11,152 @@
+                    [=](sycl::id<1> i) {
+                        ipivOut[i] = static_cast<int64_t>(ipivIn[i]);
+                    });
+
+        });
+    */
+
+    Array<compute_t<T>> B = copyArray<compute_t<T>>(b);
+    sycl::buffer<compute_t<T>> aBuf =
+        A.template getBufferWithOffset<compute_t<T>>();
+    sycl::buffer<compute_t<T>> bBuf =
+        B.template getBufferWithOffset<compute_t<T>>();
+
+    ::oneapi::mkl::lapack::getrs(getQueue(), opts, N, NRHS, aBuf, LDA, ipivBuf,
+                                 bBuf, LDB, scratchpad, scratchpad_size);
+    return B;
+}
+
+template<typename T>
+Array<T> generalSolve(const Array<T> &a, const Array<T> &b) {
+    int batches = a.dims()[2] * a.dims()[3];
+
+    dim4 aDims = a.dims();
+    dim4 bDims = b.dims();
+    int M      = aDims[0];
+    int N      = aDims[1];
+    int K      = bDims[1];
+    int MN     = std::min(M, N);
+
+    int lda     = a.strides()[1];
+    int astride = a.strides()[2];
 
-    if (OpenCLCPUOffload()) { return cpu::solveLU(A, pivot, b, options); }
+    sycl::buffer<int64_t> ipiv(MN * batches);
+    int ipivstride = MN;
 
-    int N    = A.dims()[0];
-    int NRHS = b.dims()[1];
+    int ldb     = b.strides()[1];
+    int bstride = b.strides()[2];
 
-    vector<int> ipiv(N);
-    copyData(&ipiv[0], pivot);
+    vector<int> info(batches, 0);
 
+    Array<T> A = copyArray<T>(a);
     Array<T> B = copyArray<T>(b);
 
-    const Buffer *A_buf = A.get();
-    Buffer *B_buf       = B.get();
+    std::int64_t scratchpad_size =
+        ::oneapi::mkl::lapack::getrf_batch_scratchpad_size<compute_t<T>>(
+            getQueue(), M, N, lda, astride, ipivstride, batches);
 
-    int info = 0;
-    magma_getrs_gpu<T>(MagmaNoTrans, N, NRHS, (*A_buf)(), A.getOffset(),
-                       A.strides()[1], &ipiv[0], (*B_buf)(), B.getOffset(),
-                       B.strides()[1], getQueue()(), &info);
-    return B;
-}
+    sycl::buffer<compute_t<T>> scratchpad(scratchpad_size);
 
-template<typename T>
-Array<T> generalSolve(const Array<T> &a, const Array<T> &b) {
-    ONEAPI_NOT_SUPPORTED("generalSolve Not supported");
+    sycl::buffer<compute_t<T>> aBuf =
+        A.template getBufferWithOffset<compute_t<T>>();
+    sycl::buffer<compute_t<T>> bBuf =
+        B.template getBufferWithOffset<compute_t<T>>();
+    ::oneapi::mkl::lapack::getrf_batch(getQueue(), M, N, aBuf, lda, astride,
+                                       ipiv, ipivstride, batches, scratchpad,
+                                       scratchpad_size);
 
-    // dim4 aDims = a.dims();
-    // int batchz = aDims[2];
-    // int batchw = aDims[3];
+    scratchpad_size =
+        ::oneapi::mkl::lapack::getrs_batch_scratchpad_size<compute_t<T>>(
+            getQueue(), ::oneapi::mkl::transpose::nontrans, N, K, lda, astride,
+            ipivstride, ldb, bstride, batches);
 
-    // Array<T> A = copyArray<T>(a);
-    Array<T> B = copyArray<T>(b);
+    // TODO: reuse? or single large scratchpad?
```

Comment:
You can't reuse right? the size is going to be different every time. One thing we can do is create a memory manager that allocates the buffers of the exact size, but I don't think you want to do that at this time.